File: 1086.cpp
#include<iostream>
#include<vector>
using namespace std;

struct Record {
    bool isPush;
    int value;
    Record(bool i, int v) {
        isPush = i;
        value = v;
    }
};

vector<Record> records;

struct Range {
    int left, right;
    Range(int l, int r) {
        left = l;
        right = r;
    }
    bool isValid() {
        return left >= right;
    }
    Range getLeft() {
        int pushCount = 0, popCount = 0;
        for (int i = left + 1; i < right; i++) {
            if (records[i].isPush) pushCount++;
            else popCount++;
            if (popCount > pushCount) return Range(left + 1, i);
        }
        return Range(left + 1, right);
    }
};

struct Node {
    int value, left, right;
    Node() {
        value = -1; left = -1, right = -1;
    }
    Node(int l, int r) {
        left = l;
        right = r;
    }
};

Node nodes[31];

int getTree(Range range) {
    if (range.isValid()) return -1;
    int root = records[range.left].value;
    Range lr = range.getLeft();
    Range rr = Range(lr.right + 1, range.right);
    nodes[root].left = getTree(lr);
    nodes[root].right = getTree(rr);
    return root;
}

bool flag = true;
void postTravel(int root) {
    if (nodes[root].left != -1) postTravel(nodes[root].left);
    if (nodes[root].right != -1) postTravel(nodes[root].right);
    if (flag) flag = 0;
    else cout << " ";
    cout << root;
}

int main() {
    int n;
    cin >> n;
    for (int i = 0; i < n * 2; i++) {
        string s;
        cin >> s;
        if (s == "Push") {
            int v;
            cin >> v;
            records.push_back(Record(true, v));
        } else {
            records.push_back(Record(false, -2));
        }
    }
    int root = getTree(Range(0, n * 2));
    postTravel(root);
}

File: 1086.swift

import Foundation

class Tree {
    var value: Int
    var left, right: Tree?
    init(_ value: Int) {
        self.value = value
    }
}

enum Record {
    indirect case push(Int)
    case pop
}

var records = [Record]()

struct Range {
    let left, right: Int
    var isValid: Bool {
        return left >= right
    }
    func getRange() -> (Range?, Range?) {
        var pushCount = 0, popCount = 0
        for i in left + 1..<right {
            switch records[i] {
            case .push(_):  pushCount += 1
            case .pop:      popCount += 1
            }
            if popCount > pushCount {
                return (Range(left: left + 1, right: i), Range(left: i + 1, right: right))
            }
        }
        return (nil, nil)
    }
    func getTree() -> Tree? {
        if isValid {
            return nil
        } else {
            switch records[left] {
            case .push(let value):
                let tree = Tree(value)
                let (left, right) = getRange()
                tree.left = left?.getTree()
                tree.right = right?.getTree()
                return tree
            default:
                return nil
            }
        }
    }
}

var flag = true

extension Tree {
    func postTravel() {
        left?.postTravel()
        right?.postTravel()
        flag ? flag = false : print(terminator: " ")
        print(value, terminator: "")
    }
}

let n = Int(readLine()!)!

for _ in 0 ..< n * 2 {
    let s = readLine()!
    if s.hasPrefix("Push") {
        let value = String(s[s.index(s.startIndex, offsetBy: 5)...])
        records.append(Record.push(Int(value)!))
    } else {
        records.append(Record.pop)
    }
}

Range(left: 0, right: n * 2).getTree()?.postTravel()
